packages/eventual-send/src/index.js

```diff
@@ -71,33 +68,109 @@ export function makeHandledPromise() {
         promiseToPresence.set(target, presence);
         target = parent;
       }
     } else {
-      // We propagate p and remove all other unsettled handlers
+      // We propagate p and remove all other pending handlers
       // upstream.
       // Note that everything except presences is covered here.
       while (target !== p) {
         const parent = forwardedPromiseToPromise.get(target);
         forwardedPromiseToPromise.set(target, p);
-        promiseToUnsettledHandler.delete(target);
+        promiseToPendingHandler.delete(target);
         target = parent;
       }
     }
     return target;
   }
 
-  // This special handler accepts Promises, and forwards
-  // handled Promises to their corresponding fulfilledHandler.
+  /**
+   * This special handler accepts Promises, and forwards
+   * handled Promises to their corresponding fulfilledHandler.
+   *
+   * @type {Required<EHandler<any>>}
+   */
   let forwardingHandler;
   let handle;
 
-  function HandledPromise(executor, unsettledHandler = undefined) {
-    if (new.target === undefined) {
-      throw new Error('must be invoked with "new"');
+  /**
+   * @param {string} handlerName
+   * @param {EHandler<any>} handler
+   * @param {string} operation
+   * @param {any} o
+   * @param {any[]} opArgs
+   * @param {Promise<unknown>} [returnedP]
+   * @returns {any}
+   */
+  const dispatchToHandler = (
+    handlerName,
+    handler,
+    operation,
+    o,
+    opArgs,
+    returnedP,
+  ) => {
+    let actualOp = operation;
+    const triedOperations = [operation];
+    const isSendOnly = actualOp.endsWith('SendOnly');
+    const makeResult = result => (isSendOnly ? undefined : result);
+
+    if (isSendOnly) {
+      // We don't specify the resulting promise if it is sendonly.
+      returnedP = undefined;
+    }
+
+    if (isSendOnly && typeof handler[actualOp] !== 'function') {
+      // Substitute for sendonly with the corresponding non-sendonly operation.
+      actualOp = actualOp.slice(0, -'SendOnly'.length);
+      triedOperations.push(actualOp);
+    }
+
+    // Fast path: just call the actual operation.
+    if (typeof handler[actualOp] === 'function') {
+      return makeResult(handler[actualOp](o, ...opArgs, returnedP));
     }
+
+    if (actualOp === 'applyMethod') {
+      // Compose a missing applyMethod by get followed by applyFunction.
+      const [prop, args] = opArgs;
+      const getResultP = handle(o, 'get', [prop], undefined);
+      return makeResult(handle(getResultP, 'applyFunction', [args], returnedP));
+    }
+
+    if (actualOp === 'applyFunction') {
+      actualOp = 'applyMethod';
+      triedOperations.push(actualOp);
+      if (typeof handler[actualOp] === 'function') {
+        // Downlevel a missing applyFunction to applyMethod with undefined name.
+        const [args] = opArgs;
+        return makeResult(handler[actualOp](o, undefined, [args], returnedP));
+      }
+    }
+
+    const { details: X, quote: q } = assert;
```

Comment:
Oh I see we are depending on `assert`. Fine with me! In that case, we should use it where we can to improve the code quality. We should always use its `q` rather than `JSON.stringify`. We should generally use
```js
assert(cond, X`msg`);
```
rather than
```
if (!cond) { throw TypeError(`msg-without-X`); }
```
We have too many of the latter.

We should also move initialization lines like this right below the imports at the top of the file. Having it appear nested within a function is especially weird.

Reply:
> We should also move initialization lines like this right below the imports at the top of the file. Having it appear nested within a function is especially weird.

Ah, right.  I thought that `import 'ses';` didn't leave us with a working `assert` until after `lockdown`, but I see that it does.  I'll move these things up.

Reply:
> We should generally use
> 
> ```js
> assert(cond, X`msg`);
> ```

The ones that remain are the things I didn't want to add the gathering of extra debugging information to the hot path.